File: libs/wasm-sdk/wasm-sdk-python/flow_like_wasm_sdk/context.py
from __future__ import annotations

import json
from typing import Any

from flow_like_wasm_sdk.host import HostBridge, _host
from flow_like_wasm_sdk.types import ExecutionInput, ExecutionResult, LogLevel


class Context:
    """Execution context providing typed input access, output setting, logging, and streaming."""

    def __init__(self, execution_input: ExecutionInput, host: HostBridge | None = None) -> None:
        self._input = execution_input
        self._result = ExecutionResult.ok()
        self._host = host or _host

    @classmethod
    def from_dict(cls, data: dict[str, Any], host: HostBridge | None = None) -> Context:
        return cls(ExecutionInput.from_dict(data), host)

    @classmethod
    def from_json(cls, json_str: str, host: HostBridge | None = None) -> Context:
        return cls(ExecutionInput.from_json(json_str), host)

    @property
    def node_id(self) -> str:
        return self._input.node_id

    @property
    def node_name(self) -> str:
        return self._input.node_name

    @property
    def run_id(self) -> str:
        return self._input.run_id

    @property
    def app_id(self) -> str:
        return self._input.app_id

    @property
    def board_id(self) -> str:
        return self._input.board_id

    @property
    def user_id(self) -> str:
        return self._input.user_id

    @property
    def stream_enabled(self) -> bool:
        return self._input.stream_state

    @property
    def log_level(self) -> int:
        return self._input.log_level

    def get_input(self, name: str) -> Any:
        return self._input.inputs.get(name)

    def get_string(self, name: str, default: str | None = None) -> str | None:
        val = self.get_input(name)
        if val is None:
            return default
        return str(val)

    def get_i64(self, name: str, default: int | None = None) -> int | None:
        val = self.get_input(name)
        if val is None:
            return default
        return int(val)

    def get_f64(self, name: str, default: float | None = None) -> float | None:
        val = self.get_input(name)
        if val is None:
            return default
        return float(val)

    def get_bool(self, name: str, default: bool | None = None) -> bool | None:
        val = self.get_input(name)
        if val is None:
            return default
        return bool(val)

    def require_input(self, name: str) -> Any:
        val = self.get_input(name)
        if val is None:
            raise ValueError(f"Required input '{name}' not provided")
        return val

    def set_output(self, name: str, value: Any) -> None:
        self._result.set_output(name, value)

    def activate_exec(self, pin_name: str) -> None:
        self._result.exec(pin_name)

    def set_pending(self, pending: bool) -> None:
        self._result.set_pending(pending)

    def debug(self, message: str) -> None:
        if self._input.log_level <= LogLevel.DEBUG:
            self._host.log(LogLevel.DEBUG, message)

    def info(self, message: str) -> None:
        if self._input.log_level <= LogLevel.INFO:
            self._host.log(LogLevel.INFO, message)

    def warn(self, message: str) -> None:
        if self._input.log_level <= LogLevel.WARN:
            self._host.log(LogLevel.WARN, message)

    def error(self, message: str) -> None:
        if self._input.log_level <= LogLevel.ERROR:
            self._host.log(LogLevel.ERROR, message)

    def stream_text(self, text: str) -> None:
        if self._input.stream_state:
            self._host.stream("text", text)

    def stream_json(self, data: Any) -> None:
        if self._input.stream_state:
            self._host.stream("json", json.dumps(data))

    def stream_progress(self, progress: float, message: str) -> None:
        if self._input.stream_state:
            payload = json.dumps({"progress": progress, "message": message})
            self._host.stream("progress", payload)

    def get_variable(self, name: str) -> Any:
        return self._host.get_variable(name)

    def set_variable(self, name: str, value: Any) -> bool:
        return self._host.set_variable(name, value)

    def delete_variable(self, name: str) -> None:
        self._host.delete_variable(name)

    def has_variable(self, name: str) -> bool:
        return self._host.has_variable(name)

    def cache_get(self, key: str) -> Any:
        return self._host.cache_get(key)

    def cache_set(self, key: str, value: Any) -> None:
        self._host.cache_set(key, value)

    def cache_delete(self, key: str) -> None:
        self._host.cache_delete(key)

    def cache_has(self, key: str) -> bool:
        return self._host.cache_has(key)

    def storage_dir(self, node_scoped: bool = False) -> dict | None:
        return self._host.storage_dir(node_scoped)

    def upload_dir(self) -> dict | None:
        return self._host.upload_dir()

    def cache_dir(self, node_scoped: bool = False, user_scoped: bool = False) -> dict | None:
        return self._host.cache_dir(node_scoped, user_scoped)

    def user_dir(self, node_scoped: bool = False) -> dict | None:
        return self._host.user_dir(node_scoped)

    def storage_read(self, flow_path: dict) -> bytes | None:
        return self._host.storage_read(flow_path)

    def storage_write(self, flow_path: dict, data: bytes) -> bool:
        return self._host.storage_write(flow_path, data)

    def storage_list(self, flow_path: dict) -> list[dict] | None:
        return self._host.storage_list(flow_path)

    def embed_text(self, bit: dict, texts: list[str]) -> list[list[float]] | None:
        return self._host.embed_text(bit, texts)

    def http_request(self, method: int, url: str, headers: dict[str, str] | None = None, body: bytes | None = None) -> dict | None:
        result = self._host.http_request(method, url, json.dumps(headers or {}), body)
        if result is None:
            return None
        return json.loads(result)

    def http_get(self, url: str, headers: dict[str, str] | None = None) -> dict | None:
        return self.http_request(0, url, headers)

    def http_post(self, url: str, body: bytes | None = None, headers: dict[str, str] | None = None) -> dict | None:
        return self.http_request(1, url, headers, body)

    def get_oauth_token(self, provider: str) -> dict | None:
        return self._host.get_oauth_token(provider)

    def has_oauth_token(self, provider: str) -> bool:
        return self._host.has_oauth_token(provider)

    def success(self) -> ExecutionResult:
        self._result.exec("exec_out")
        return self._result

    def fail(self, error: str) -> ExecutionResult:
        self._result.error = error
        return self._result

    def finish(self) -> ExecutionResult:
        return self._result

File: libs/wasm-sdk/wasm-sdk-csharp/Context.cs
namespace FlowLike.Wasm.Sdk;

public class Context
{
    private readonly ExecutionInput _input;
    private readonly ExecutionResult _result;
    private readonly HostBridge _host;

    public Context(ExecutionInput input, HostBridge? host = null)
    {
        _input = input;
        _result = ExecutionResult.Ok();
        _host = host ?? Host.Current;
    }

    public static Context FromJson(string json, HostBridge? host = null) =>
        new(ExecutionInput.FromJson(json), host);

    public string NodeId => _input.NodeId;
    public string NodeName => _input.NodeName;
    public string RunId => _input.RunId;
    public string AppId => _input.AppId;
    public string BoardId => _input.BoardId;
    public string UserId => _input.UserId;
    public bool StreamEnabled => _input.StreamState;
    public int LogLevelValue => _input.LogLevelValue;

    public object? GetInput(string name) =>
        _input.Inputs.TryGetValue(name, out var val) ? val : null;

    public string? GetString(string name, string? defaultValue = null)
    {
        var val = GetInput(name);
        return val is not null ? val.ToString() : defaultValue;
    }

    public long? GetI64(string name, long? defaultValue = null)
    {
        var val = GetInput(name);
        if (val is null) return defaultValue;
        return Convert.ToInt64(val);
    }

    public double? GetF64(string name, double? defaultValue = null)
    {
        var val = GetInput(name);
        if (val is null) return defaultValue;
        return Convert.ToDouble(val);
    }

    public bool? GetBool(string name, bool? defaultValue = null)
    {
        var val = GetInput(name);
        if (val is null) return defaultValue;
        return Convert.ToBoolean(val);
    }

    public object RequireInput(string name) =>
        GetInput(name) ?? throw new InvalidOperationException($"Required input '{name}' not provided");

    public void SetOutput(string name, object? value) => _result.SetOutput(name, value);
    public void ActivateExec(string pinName) => _result.Exec(pinName);
    public void SetPending(bool pending) => _result.SetPending(pending);

    public void Debug(string message) { if (_input.LogLevelValue <= (int)LogLevel.Debug) _host.Log((int)LogLevel.Debug, message); }
    public void Info(string message) { if (_input.LogLevelValue <= (int)LogLevel.Info) _host.Log((int)LogLevel.Info, message); }
    public void Warn(string message) { if (_input.LogLevelValue <= (int)LogLevel.Warn) _host.Log((int)LogLevel.Warn, message); }
    public void Error(string message) { if (_input.LogLevelValue <= (int)LogLevel.Error) _host.Log((int)LogLevel.Error, message); }

    public void StreamText(string text)
    {
        if (_input.StreamState) _host.Stream("text", text);
    }

    public void StreamJson(object data)
    {
        if (_input.StreamState) _host.Stream("json", Json.Serialize(data));
    }

    public void StreamProgress(double progress, string message)
    {
        if (_input.StreamState)
        {
            var payload = Json.Serialize(new Dictionary<string, object> { ["progress"] = progress, ["message"] = message });
            _host.Stream("progress", payload);
        }
    }

    public object? GetVariable(string name) => _host.GetVariable(name);
    public bool SetVariable(string name, object? value) => _host.SetVariable(name, value);
    public void DeleteVariable(string name) => _host.DeleteVariable(name);
    public bool HasVariable(string name) => _host.HasVariable(name);

    public object? CacheGet(string key) => _host.CacheGet(key);
    public void CacheSet(string key, object? value) => _host.CacheSet(key, value);
    public void CacheDelete(string key) => _host.CacheDelete(key);
    public bool CacheHas(string key) => _host.CacheHas(key);

    public Dictionary<string, object?>? StorageDir(bool nodeScoped = false) => _host.StorageDir(nodeScoped);
    public Dictionary<string, object?>? UploadDir() => _host.UploadDir();
    public Dictionary<string, object?>? CacheDir(bool nodeScoped = false, bool userScoped = false) => _host.CacheDir(nodeScoped, userScoped);
    public Dictionary<string, object?>? UserDir(bool nodeScoped = false) => _host.UserDir(nodeScoped);

    public byte[]? StorageRead(Dictionary<string, object?> flowPath) => _host.StorageRead(flowPath);
    public bool StorageWrite(Dictionary<string, object?> flowPath, byte[] data) => _host.StorageWrite(flowPath, data);
    public List<Dictionary<string, object?>>? StorageList(Dictionary<string, object?> flowPath) => _host.StorageList(flowPath);

    public Dictionary<string, object?>? HttpRequest(byte method, string url, Dictionary<string, string>? headers = null, byte[]? body = null)
    {
        var result = _host.HttpRequest(method, url, Json.Serialize(headers ?? new Dictionary<string, string>()), body);
        if (result == null) return null;
        return Json.Deserialize<Dictionary<string, object?>>(result);
    }

    public Dictionary<string, object?>? HttpGet(string url, Dictionary<string, string>? headers = null) =>
        HttpRequest(0, url, headers);

    public Dictionary<string, object?>? HttpPost(string url, byte[]? body = null, Dictionary<string, string>? headers = null) =>
        HttpRequest(1, url, headers, body);

    public ExecutionResult Success()
    {
        _result.Exec("exec_out");
        return _result;
    }

    public ExecutionResult Fail(string error)
    {
        _result.Error = error;
        return _result;
    }

    public ExecutionResult Finish() => _result;
}
